src/software/jetson_nano/services/motor.cpp

```diff
@@ -102,54 +111,223 @@ MotorService::MotorService(const RobotConstants_t& robot_constants,
-
+                    motor.direct_per_wheel_control().front_right_wheel_rpm() *
+                    robot_constants_.wheel_rotations_per_motor_rotation));
+            tmc4671_setTargetVelocity(
+                BACK_LEFT_MOTOR_CHIP_SELECT,
+                static_cast<int>(motor.direct_per_wheel_control().back_left_wheel_rpm() *
+                                 robot_constants_.wheel_rotations_per_motor_rotation));
+            tmc4671_setTargetVelocity(
+                BACK_RIGHT_MOTOR_CHIP_SELECT,
+                static_cast<int>(motor.direct_per_wheel_control().back_right_wheel_rpm() *
+                                 robot_constants_.wheel_rotations_per_motor_rotation));
             break;
         }
-        case TbotsProto::DirectControlPrimitive::WheelControlCase::kDirectVelocityControl:
+        case TbotsProto::MotorControl::DriveControlCase::kDirectVelocityControl:
         {
-            // TODO (#2335) convert local velocity to per-wheel velocity
-            // using http://robocup.mi.fu-berlin.de/buch/omnidrive.pdf and then
-            // communicate velocities to trinamic.
+            EuclideanSpace_t target_euclidean_velocity = {
+                motor.direct_velocity_control().velocity().x_component_meters(),
+                motor.direct_velocity_control().velocity().y_component_meters(),
+                motor.direct_velocity_control().angular_velocity().radians_per_second(),
+            };
+
+            // This is a linear transformation, we don't need to convert to/from
+            // RPM to MPS
+            WheelSpace_t target_speeds = euclidean_to_four_wheel.getTargetWheelSpeeds(
+                target_euclidean_velocity, current_wheel_speeds);
+
+            tmc4671_setTargetVelocity(FRONT_RIGHT_MOTOR_CHIP_SELECT,
+                                      static_cast<int>(target_speeds[0]));
+            tmc4671_setTargetVelocity(FRONT_LEFT_MOTOR_CHIP_SELECT,
+                                      static_cast<int>(target_speeds[1]));
+            tmc4671_setTargetVelocity(BACK_LEFT_MOTOR_CHIP_SELECT,
+                                      static_cast<int>(target_speeds[2]));
+            tmc4671_setTargetVelocity(BACK_RIGHT_MOTOR_CHIP_SELECT,
+                                      static_cast<int>(target_speeds[3]));
+
             break;
         }
-        case TbotsProto::DirectControlPrimitive::WheelControlCase::WHEEL_CONTROL_NOT_SET:
+
+        case TbotsProto::MotorControl::DriveControlCase::DRIVE_CONTROL_NOT_SET:
         {
-            LOG(WARNING) << "Motor service polled with an empty DirectControlPrimitive ";
+            LOG(WARNING) << "Motor service polled with an empty DirectControlPrimitive";
             break;
         }
     }
 
-    return std::make_unique<TbotsProto::DriveUnitStatus>();
+    // Toggle Hearbeat
+    heartbeat_gpio.setValue((heartbeat_state = !heartbeat_state) ? GpioState::HIGH
+                                                                 : GpioState::LOW);
```

Comment:
usually we don't use ternary operators

Reply:
lol imo we gotta change that rule, i think short ternary operators are fairly readable, its the ones that go on for a while that are annoying

I've changed it to

~~~
  316     // Toggle Hearbeat                                                                                                                                   
~   1     if (heartbeat_state == 1)                                                                                                                            
~   2     {                                                                                                                                                    
+   3         heartbeat_gpio.setValue(GpioState::LOW);                                                                                                         
+   4         heartbeat_state = 0;                                                                                                                             
+   5     }                                                                                                                                                    
+   6     else                                                                                                                                                 
+   7     {                                                                                                                                                    
+   8         heartbeat_gpio.setValue(GpioState::HIGH);                                                                                                        
+   9         heartbeat_state = 1;                                                                                                                             
+  10     }                           
~~~

Reply:
well if you have a good reason to use ternary I will not block the pr from being merged

Reply:
its 1 line instead of 10 but that's not a good enough reason

I've already changed it back to if/elses